src/kudu/consensus/consensus_queue.cc

```diff
@@ -883,6 +887,97 @@ void PeerMessageQueue::AdvanceQueueWatermark(const char* type,
   }
 }
 
+void PeerMessageQueue::AdvanceMajorityReplicatedWatermark(
+    int64_t* watermark, const OpId& replicated_before,
+    const OpId& replicated_after,
+    const std::map<std::string, int>& data_commit_quorum,
+    ReplicaTypes replica_types, const TrackedPeer* who_caused) {
+  CHECK(watermark);
+  CHECK(who_caused);
+
+  std::string type = "majority_replicated";
+  if (VLOG_IS_ON(2)) {
+    VLOG_WITH_PREFIX_UNLOCKED(2) << "Updating " << type << " watermark: "
+        << "Peer (" << who_caused->ToString() << ") changed from "
+        << replicated_before << " to " << replicated_after << ". "
+                                 << "Current value: " << *watermark;
+  }
+
+  // For each region, we compute a vector of indexes that were replicated.
+  std::map<std::string, vector<int64_t> > watermarks_by_region;
+  for (const PeersMap::value_type& peer : peers_map_) {
+    if (replica_types == VOTER_REPLICAS &&
+        peer.second->peer_pb.member_type() != RaftPeerPB::VOTER) {
+      continue;
+    }
+    // Refer to the comment in AdvanceQueueWatermark method for why only
+    // successful last exchanges are considered.
+    if (peer.second->last_exchange_status == PeerStatus::OK) {
+      const string& peer_region = peer.second->peer_pb.attrs().region();
+      std::vector<int64_t>& regional_watermarks = LookupOrInsert(
+          &watermarks_by_region, peer_region, std::vector<int64_t>());
+      regional_watermarks.push_back(peer.second->last_received.index());
+    }
+  }
+
+  // Vector of maximum indexes which satisfy commit requirements in each region.
+  std::vector<int64_t> commit_index_regional_level;
+  for (const std::pair<std::string, int>& regional_commit_requirement :
+      data_commit_quorum) {
+    std::map<std::string, std::vector<int64_t> >::iterator it =
+        watermarks_by_region.find(regional_commit_requirement.first);
+    if (it == watermarks_by_region.end() ||
+        it->second.size() < regional_commit_requirement.second) {
+      VLOG_WITH_PREFIX_UNLOCKED(3)
+          << "Watermarks size: "
+          << ((it == watermarks_by_region.end()) ? 0 : it->second.size())
+          << ", Num peers required: " << regional_commit_requirement.second
+          << ", Region: " << regional_commit_requirement.first;
+      return;
+    }
+    std::sort(it->second.begin(), it->second.end());
+    commit_index_regional_level.push_back(
```

Comment:
needs a comment that you are looking for the J'th entry where J is the size of the quorum